Automattic/wp-calypso
# Correct the storage dropdown list data

Related to #97221 

## Proposed Changes
This PR is based on https://github.com/Automattic/wp-calypso/pull/97794, aiming to fix one of the issues described there:

> Another known issue is that the total spaces shown in the storage dropdown are wrong for a site owning a storage add-on. This will also be fixed as a follow-up.

The bug is caused by that the code computing for the total storage doesn't take into account that the storage add-ons are actually tiered products, so it'd add up the attempted storage add-on, which should be replacing instead. Thus, the bug will only happen when accessing `/plans` using a site having a storage add-on.

e.g. A Business site with 100G add-on
Before:
![CleanShot 2024-12-31 at 14 22 08@2x](https://github.com/user-attachments/assets/5d689903-bd02-4ee1-b6af-3b6afc77cdd8)

After:
![CleanShot 2024-12-31 at 14 19 42@2x](https://github.com/user-attachments/assets/7c03ad06-7bad-4ac8-a6c0-510861a51511)

Along side the bug fix, several refactors are introduced to simplify the code further:

* Introduce the hook, `use-plan-storage`, `use-purchased-add-on`, and `use-storage-string`. In general the logic is to list the total storage space as `(storage provided by the plan + storage provided by the add-on)`. Thus, this new set of hooks are introduced to derive these info.
* `use-available-storage-options`, `use-storage-string-from-feature` are deprecated.

## Why are these changes being made?


It's crucial since the bug is revealed by the extended storage add-ons. 

## Testing Instructions



When testing on anywhere other than the local calypso, please remember to add `?flags=upgrades/storage-add-on-v2`

* Logging in to a site having no storage add-on, confirm that the storage dropdown in `/plans` enlists all the options correctly for the Business plan and the Commerce plan. For Free/Personal/Premium, the storage should be shown as labels as expected.
* Log in to a site having a storage add-on and do the same.

## Pre-merge Checklist
